jax/_src/api_util.py

```diff
@@ -587,3 +583,79 @@ def _np_scalar_abstractify(x: np.generic) -> ShapedArray:
 # By default it does nothing, but it can be monkey-patched to do other things.
 def api_hook(fun, tag: str):
   return fun
+
+
+class TracingDebugInfo(NamedTuple):
+  traced_for: str             # e.g. 'jit', 'scan', etc
+  func_src_info: str          # e.g. f'{fun.__name__} at {filename}:{lineno}'
+  arg_names: Tuple[str, ...]  # e.g. ('args[0]', ... )
+
+def debug_info(traced_for: str, fun: Callable, args: Tuple[Any],
+               kwargs: Dict[str, Any], static_argnums: Tuple[int, ...],
+               static_argnames: Tuple[str, ...]) -> Optional[TracingDebugInfo]:
+  src = fun_sourceinfo(fun)
+  arg_names = _arg_names(fun, args, kwargs, static_argnums, static_argnames)
+  if src is None or arg_names is None: return None
+  return TracingDebugInfo(traced_for, src, arg_names)
+
+def fun_sourceinfo(fun: Callable) -> Optional[str]:
+  while isinstance(fun, partial):
+    fun = fun.func
+  fun = inspect.unwrap(fun)
+  try:
+    filename = fun.__code__.co_filename
+    lineno = fun.__code__.co_firstlineno
+    return f"{fun.__name__} at {filename}:{lineno}"
+  except AttributeError:
+    return None
+
+def _arg_names(fn, args, kwargs, static_argnums, static_argnames,
+               ) -> Optional[Tuple[str, ...]]:
+  static = object()
+  static_argnums_ = _ensure_inbounds(True, len(args), static_argnums)
+  static_argnames_ = set(static_argnames)
+  args_ = [static if i in static_argnums_ else x for i, x in enumerate(args)]
+  kwargs = {k:static if k in static_argnames_ else x for k, x in kwargs.items()}
+  try:
+    ba = inspect.signature(fn).bind(*args_, **kwargs)
+  except (ValueError, TypeError):
+    return None
+  return tuple(f'{name}{keystr(path)}' for name, x in ba.arguments.items()
+               for path, l in _generate_key_paths(x) if l is not static)
+
+def result_paths(f): return _result_paths(f, None)
+
+@lu.transformation_with_aux
+def _result_paths(_, *args, **kwargs):
+  ans = yield args, kwargs
+  yield ans, [keystr(path) for path, _ in _generate_key_paths(ans)]
+
+def jaxpr_debug_info(jaxpr: core.Jaxpr, trace_debug: Optional[TracingDebugInfo],
+                     result_paths: Optional[Tuple[Optional[str], ...]]
+                     ) -> core.Jaxpr:
+  if trace_debug is not None and result_paths is not None:
+    debug_info = core.JaxprDebugInfo(
+        trace_debug.traced_for, trace_debug.func_src_info,
+        trace_debug.arg_names, result_paths)
+  else:
+    debug_info = None
+  return jaxpr.replace(debug_info=debug_info) if debug_info else jaxpr
+
+def debug_info_final(traced_for, f, args, kwargs, static_argnums, static_argnames):
+  d = debug_info(traced_for, f.f, args, kwargs, static_argnums, static_argnames)
+  f, _ = _result_paths(f, d)
+  return f
+
+def get_debug_info_final(
```

Comment:
I've been staring at this function for 5 minutes trying to figure why it exists. (A comment would save me at least 5 minutes).

Why are we looking through the stores of a lu.WrappedFun directly as opposed to calling the thunk that would return the values in the usual way?

Reply:
Uh oh, looks like I _did_ delete it, but somehow didn't push. Luckily that means with a push now it's deleted!